torch/distributed/_composable/fsdp/fully_shard.py

```diff
@@ -0,0 +1,340 @@
+        needed. This method is *not* recursive.
+        """
+        if (state := _get_module_fsdp_state(cast(nn.Module, self))) is None or (
+            (fsdp_param_group := state._fsdp_param_group) is None
+        ):
+            return  # no-op
+        fsdp_param_group._reshard()
+
+    def set_requires_gradient_sync(
+        self, requires_gradient_sync: bool, recurse: bool = True
+    ) -> None:
+        """
+        Sets if the module should sync gradients. This can be used to implement
+        gradient accumulation without communication. For HSDP, this controls
+        both reduce-scatter and all-reduce together.
+
+        Args:
+            requires_gradient_sync (bool): Whether to reduce gradients for the
+                module's parameters.
+            recurse (bool): Whether to set for all submodules or just the
+                passed-in module.
+        """
+        module = cast(nn.Module, self)
+        states = (
+            [_get_module_fsdp_state(module)]
+            if not recurse
+            else [_get_module_fsdp_state(submodule) for submodule in module.modules()]
+        )
+        for state in states:
+            if state:
+                state._set_reduce_scatter_grads(requires_gradient_sync)
+                state._set_all_reduce_grads(requires_gradient_sync)
+
+    def set_requires_all_reduce(self, requires_all_reduce: bool, recurse: bool = True):
+        """
+        Sets if the module should all-reduce gradients. This can be used to
+        implement gradient accumulation with only reduce-scatter but not
+        all-reduce for HSDP.
+        """
+        module = cast(nn.Module, self)
+        states = (
+            [_get_module_fsdp_state(module)]
+            if not recurse
+            else [_get_module_fsdp_state(submodule) for submodule in module.modules()]
+        )
+        for state in states:
+            if state:
+                state._set_all_reduce_grads(requires_all_reduce)
+
+    # TODO: Expose a method for manually waiting on the gradient sync.
+    def set_wait_for_gradient_sync(self, wait_for_gradient_sync: bool):
+        """
+        Sets if the root module should wait for the gradient synchronization
+        (i.e. reduce-scatter and/or all-reduce) to finish at the end of
+        backward. This can be disabled for some pipeline-parallel schedules to
+        have more opportunity for communication/computation overlap. This is a
+        no-op if called on a non-root module.
+        """
+        module = cast(nn.Module, self)
+        state = _get_module_fsdp_state(module)
+        if state is None or not state._is_root:
+            raise AssertionError(f"Expects to be called on a root FSDP, not {module}")
+        state._set_wait_for_gradient_sync(wait_for_gradient_sync)
+
+    # We can properly allow `_apply` once we have: https://github.com/pytorch/pytorch/issues/113045
+    def _apply(self, *args: Any, **kwargs):
```

Comment:
Eventually, we do need/want to properly support `_apply`. One thing to watch out for there will be to make sure that CPU offloaded parameters are _not_ moved to GPU.

We can discuss where to draw the line between "users should be careful themselves if enabling CPU offloading" vs. "FSDP should be add extra logic to ensure offloaded parameters are not moved".